### DISCUSSION TITLE
Using the -k option

### DISCUSSION THREAD

**Discussion by JReally** [rustdesk/rustdesk#973] [Q&A]
I'm using self-hosting with docker (Linux Mint)
Connections via LAN and WAN work well and are encrypted

The -k option to help restrict the user of my server sounded attractive so
I reset the docker installation with -k and the appropriate key form ID_ED....pub
1. trying to setup a connection with no key specified in ID/Relay was blocked
      as expected so users without the key cannot use the server - that's good.
2. Connections when the key was specified succeeded but were always unencrypted

Not sure what I missed.... obviously I still want encrypted connections.

I have reset the docker installation without the -k option and everything works as before
except now someone without the key can still use the server.

Does anyone have an idea what I missed?

Thanks,
  JR

**Comment by sonoracomm** [comment]
I am an absolute newbie with RustDesk...

Here is the doc for what you are asking. https://rustdesk.com/docs/en/self-host/install/#key

I'm using encryption, but I didn't use "-k" to force it.

It looks like you might need "-k _" (underscore) to force encryption?

Sorry, that's all I know.

G

p.s. To the RustDesk creators/team, thank you very much for this great software!

**Reply by JReally** [reply]
Perfect solution... I was entering -k followed by the actual key.
I simply missed the _

Entering -k _  works to make the server is relatively secure from unauthorized usage.

Thanks sonoracomm and RustDesk developers

JR

**Comment by rustdesk** [comment]
https://www.youtube.com/watch?v=EeFqj23jxMk This video covers `-k` very well

**Comment by Benny7587** [comment]
I tried using the -k _ option but ran into a weird issue. I use rustdesk for remote support and when a remote client downloads the rustdesk application from my server for a second time and a previous version is still in his download directory, the key is changing because Windows appends a version number (2) or whatever version to the filename. So when i use -k _ the connection to the server is okay but the connection between the clients is not established. Is there any way to solve this i could imagine using a closing tag behind the key value so the server knows where to stop reading the key value and not taking the version (#) into account.

**Reply by azaloum90** [reply]
The solution to this is to wrap the entire file name in commas:

-- If naming a custom .exe file for use with the encryption keys, please name it like this (e.g. DO NOT remove the commas, just fill in your key): 

`rustdesk-host=connect.domain.com,https://rustdesk-host=connect.domain.com,key=PUBLIC_KEY_HERE,.exe`

The last comma before the .exe is the part to focus on.  Modern browsers will add the (1) *after* the last comma, which should ensure that the key string is not tampered with.  Example below:

![Screenshot 2023-11-08 103611](https://github.com/rustdesk/rustdesk/assets/8965127/a660a44b-8fd2-49cf-a1cc-ac56bff2ea12)

**Comment by azaloum90** [comment]
Hi, the correct method to do this is by specifying the private key in your Docker Compose file, as shown below:

`version: '3.5'
services:
  hbbs:
    container_name: hbbs
    ports:
      - 21114:21114
      - 21115:21115
      - 21116:21116
      - 21116:21116/udp
      - 21118:21118
    image: rustdesk/rustdesk-server:latest
    **command: hbbs -r connect.domain.com:21117 --key PRIVATE_KEY_HER**E
    volumes:
      - /mnt/docker/rustdesk/hbbs:/root
    networks:
      - net_rustdesk
    depends_on:
      - hbbr
    restart: unless-stopped

  hbbr:
    container_name: hbbr
    ports:
      - 21117:21117
      - 21119:21119
    image: rustdesk/rustdesk-server:latest
    **command: hbbr --key PRIVATE_KEY_HERE**
    volumes:
      - /mnt/docker/rustdesk/hbbr:/root
    networks:
      - net_rustdesk
    restart: unless-stopped
`

After that, you should configure RustDesk like this (if link is broken, please put PUBLIC KEY in the "key" field of the RustDesk application.  DO NOT put the private key in this field): https://imgur.com/a/60sA99x

If naming a custom .exe file for use with the encryption keys, please name it like this (do not remove the commas):
rustdesk-host=connect.domain.com,key=PUBLIC_KEY_HERE,.exe

The usual key issues still apply, so be sure to follow the instructions here for creating and copying the keys from the RustDesk instance: https://rustdesk.com/docs/en/self-host/rustdesk-server-oss/install/#key
